Multiple Project updates

matrei opened the pull request:
This PR includes several updates across multiple projects. Please refer to the individual commit messages for detailed descriptions of the changes.

https://github.com/grails/grails-gradle-plugin/pull/394 is needed for publishing to work.

matrei committed:
ci: improve GitHub workflows for clarity and stability

- Add descriptive job names for better output readability.
- Use explicit runner version: `ubuntu-24.04`
- Update step names with emojis for clarity.
- Replace `GRADLE_ENTERPRISE_ACCESS_KEY` with `DEVELOCITY_ACCESS_KEY`.
- Remove unused remote build cache authentication variables.
- Add `GITHUB_MAVEN_PASSWORD` for GitHub packages staging repos where missing.
- Pin version of third-party action: `ben-uk/workflow-dispatch`.
- Add Gradle `--continue` to output all potential errors.
- Add Gradle `--no-build-cache` during publishing to avoid using stale cache issues.
- Grant `packages: read` permission for GitHub packages (e.g., Grails staging repos).
- Simplify `release-notes.yml` by removing unnecessary config checks.

matrei committed:
build: switch to Gradle's `test-report-aggregation` Gradle Plugin

- Replace custom aggregate test report implementation.
- Use `test-report-aggregation` plugin for a simpler and more maintainable solution.
- Reference: [Test Report Aggregation Plugin Documentation](https://docs.gradle.org/current/userguide/test_report_aggregation_plugin.html)

matrei committed:
build: remove unused `idea` Gradle Plugin

The `idea` plugin is no longer required and has been removed to simplify the build script.

matrei committed:
build(deps): upgrade `common-custom-user-data-gradle-plugin` to `2.1`

- Bump version from `2.0.2` to `2.1`.
- [Release Notes](https://github.com/gradle/common-custom-user-data-gradle-plugin/releases/tag/v2.1)

matrei committed:
build: set correct tag for Develocity build scans

- Change incorrect tag `grails-gsp` to `grails-data-mapping`.

matrei committed:
build: remove unused `groovyVersion` property

- Groovy version is now sourced from `grails-bom`, making the property redundant.

matrei committed:
build: remove unnecessary snapshot guard

- All projects now use a unified version sourced from `projectVersion`, rendering the guard obsolete.

matrei committed:
build: correct typo in developer ID

- Fixed a typo in Puneet's developer ID

matrei committed:
build: optimize and clean up build configuration

- Use `.configureEach` where applicable for lazy configuration.
- Standardize string usage with single quotes.
- Use lifecycle logger instead of `System.out.println` for better logging.
- Address various Gradle deprecations.

matrei committed:
build: add test event logging

- Configure logging for events: passed, skipped and failed.

matrei committed:
build: add `clean` task for root project

- Define a `clean` task in the root project to simplify build cleanup.

matrei committed:
build: cleanup dependencies for better stability and clarity

- Explicitly add all direct dependencies to enhance stability (reduce reliance on transitive dependencies).
- Set the correct scope for project dependencies.
- Optionally exclude unnecessary transitive API dependencies to ensure accurate testing of dependency chains.
- Add comments to clarify why and in which scope each dependency is used.

matrei committed:
chore: remove unused imports

matrei committed:
refactor: remove unused method in test

- Clean up test code by removing an unused method, eliminating the dependency on `jakarta.annotation:jakarta.annotation-api` in this class.

matrei committed:
fix: replace usage of deprecated `BeanUtils.instantiate` method

- Update usages of `BeanUtils.instantiate` to the non-deprecated `BeanUtils.instantiateClass`.

matrei committed:
refactor: replace deprecated `AbstractGlobalExtension` with `IGlobalExtension`

- Refactor code to implement the `IGlobalExtension` interface, replacing the deprecated `AbstractGlobalExtension` class.

matrei committed:
test: replace `@Ignore` with `@PendingFeature`

- Use `@PendingFeature` so that the build fails if the test unexpectedly succeeds, providing an alert that the test passes.

matrei committed:
test: remove `@Ignore` from passing `HasManyDefaultMappedBySpec` test

- The `HasManyDefaultMappedBySpec` test now passes, so the `@Ignore` annotation is no longer needed.

matrei committed:
test: replace deprecated `@RunWith` with `@Suite`

- Update tests to use the `@Suite` annotation, replacing the deprecated `@RunWith`.

matrei committed:
refactor: replace deprecated `AccessControlException` with `SecurityException`

- Update code to use `SecurityException` instead of the deprecated `AccessControlException`.

matrei committed:
chore: update license headers for changes files

- Update year and set copyright holder to "the original author or authors."
- Change the license link to use `https`.

matrei committed:
ci: fix typo in step name

matrei committed:
build: fix building with Groovy snapshots

- Resolve issues with building using a Groovy snapshot version.
- Explicitly override the Groovy version in the build when requested, as it is now sourced from grails-bom instead of a property.

matrei committed:
build: fix typo in `test-report/build.gradle`

- Remove space

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -8,83 +8,91 @@ on:
-          java-version: 17
-      - name: Setup Gradle
+          distribution: 'liberica'
+          java-version: '17'
```
Given this is data mapping, should we have a matrix for 17, 21 at a minimum?

matrei replied:
Good catch!

jdaugherty commented on gradle.properties:
```diff
@@ -7,7 +7,6 @@ commonsValidatorVersion=1.9.0
 hibernateVersion=5.6.15.Final
 jakartaElVersion=4.0.0
 jakartaElGlassfishImplVersion=4.0.2
-groovyVersion=4.0.24
```
Thank you for removing this.  One of the changes we made in the bom was switching to 'groovy.version' instead of `groovyVersion` because spring uses that property and we don't want to define both properties (we can't without defining overrides).  Just wanted to mention this as a heads-up since I'm not sure we've widely advertised this yet.

jdaugherty replied:
FYI: This was a good reminder so I updated https://github.com/grails/grails-core/blob/7.0.x/UPGRADE7.md to include this warning.

jdaugherty commented on gradle/publish-config.gradle:
```diff
@@ -1,17 +1,17 @@
-ext.set('signing.keyId', project.findProperty('signing.keyId') ?: System.getenv('SIGNING_KEY'))
-ext.set('signing.password', project.findProperty('signing.password') ?: System.getenv('SIGNING_PASSPHRASE'))
+ext.set('signing.keyId', System.getenv('SIGNING_KEY'))
+ext.set('signing.password', System.getenv('SIGNING_PASSPHRASE'))
```
Should we just adopt the grails publish gradle plugin?

matrei replied:
I have refactored to use `org.grails.grails-publish`. However, I needed to change how BOM versions are resolved in https://github.com/grails/grails-gradle-plugin/pull/394, so we need to merge that first.

jdaugherty replied:
I approved that PR after discussion,  thank you!

jdaugherty commented on grails-datastore-core/build.gradle:
```diff
@@ -7,35 +7,59 @@ version = projectVersion
-    testImplementation "org.springframework:spring-context"
-    testImplementation "org.springframework:spring-web"
-
-    testImplementation "org.apache.groovy:groovy-test-junit5"
```
There are junit 5 tests that make use of groovy-test-case support.  I don't think you can remove this library.

See GormMappingSyntaxTests

matrei replied:
Good catch!

jdaugherty commented on grails-datastore-gorm-async/build.gradle:
```diff
@@ -7,21 +7,36 @@ version = projectVersion
 
-    testImplementation "org.junit.jupiter:junit-jupiter-api"
-    testImplementation "org.spockframework:spock-core"
+    compileOnly project(':grails-datastore-async'), {
```
Doesn't this need to be async to ensure the transformation is on the classpath for downstream projects that use this project?

matrei replied:
Good catch! I have changed to `compileOnlyApi`.

jdaugherty commented on grails-datastore-gorm-tck/build.gradle:
```diff
@@ -7,26 +7,86 @@ version = projectVersion
-    runtimeOnly "org.apache.groovy:groovy-dateutil"
-    runtimeOnly "org.junit.jupiter:junit-jupiter-engine"
+    runtimeOnly 'org.apache.groovy:groovy-dateutil', {
+        // Groovy Date Utils Extensions are used in the tests
```
If it's used in tests, why aren't we changing this to a test scope?

matrei replied:
I don't think the TCK project works this way. The TCK project does not have any test sources. The main sources are the tests. Then the consuming projects are copying the TCK main classes into their test classes output directory and adds the TCK project as a dependency to their `testImplementation` configuration.

jdaugherty replied:
I keep forgetting this is the tck project, yeah, it makes sense.

jdaugherty commented on grails-gorm-testing-support/build.gradle:
```diff
@@ -16,13 +16,106 @@ ext {
     implementation platform("org.grails:grails-bom:$grailsVersion")
-    implementation 'org.apache.groovy:groovy'
 
-    api 'org.grails:grails-plugin-domain-class'
```
Isn't this still required otherwise people will not have it included in their grails apps?  This is what has the DomainClassTraitInjector and I don't see this included anywhere else.

matrei replied:
No, I do not think so. `grails-gorm-testing-support` does not depend on `grails-plugin-domain-class`. Applications using domain classes will typically pull `grails-plugin-domain-class` directly or transitively through another dependency. I do not think It is the responsibility of `grails-gorm-testing-support` to provide this dependency to Grails applications.

jdaugherty replied:
I missed that this was the testing support include.  This makes sense to remove.

jdaugherty commented on test-report/build.gradle:
```diff
@@ -0,0 +1,27 @@
+reporting {
+    reports {
+        testAggregateTestReport(AggregateTestReport) {
+            testType = TestSuiteType.UNIT_TEST
```
This is great!  I didn't know this was built in.  But doesn't this need to be for all test types?

matrei replied:
To my knowledge, there are no other test types used in this project.

jdaugherty replied:
So `integration-test` such as ServiceInjectionSpec are considered unit tests in Gradle's view?

jdaugherty replied:
Looks like it is because we don't make use of the jvm-test-suite plugin:  https://docs.gradle.org/current/userguide/jvm_test_suite_plugin.html 

Maybe in the future we should convert integration tests to integration tests in gradle's view point?  We could remove code like this: 

    dependencies {
      integrationTestImplementation sourceSets.main.output
      integrationTestImplementation sourceSets.test.output
    }

where we define the integration tests.  I know we define it in the grails-core in 'integration-test.gradle'.  Which looks like we implement a gradle plugin to provide this to other projects:  IntegrationTestGradlePlugin.  Do we even need that plugin anymore?  Maybe that should be using the JVM test suite one?

jdaugherty replied:
to be clear, I don't think we need a change here for this PR.  But we should talk about how to adopt built-in gradle functions instead of writing our own gradle code.  This comment was mostly me thinking out loud on next steps.  Let me know what you think!

matrei committed:
test: use `@PendingFeatureIf` for selective test execution

Previously, some tests in the TCK were marked with `@Ignored`.
This commit replaces `@Ignored` with `@PendingFeatureIf`, using system property conditions to enable
selective execution. This approach ensures tests are only skipped when necessary, improving test coverage
and flexibility.

matrei committed:
ci: add Java version matrix with 17 and 21

Update the CI configuration to include a build matrix for Java versions 17 and 21.
This ensures compatibility and proper testing across multiple supported Java versions.

matrei committed:
build: restore `groovy-test-junit5` for JUnit 5 compatibility

Re-add the `groovy-test-junit5` dependency to ensure JUnit 5 tests execute correctly.

matrei committed:
build: use `compileOnlyApi` to expose AST transforms to downstream compileClasspath

Switch from `compileOnly` to `compileOnlyApi` to ensure that the `DelegateAsync` AST transformation classes are available
on the downstream `compileClasspath`.

matrei committed:
test: unignore passing test in `WhereMethodSpec`

The test `WhereMethodSpec."Test rlike query"` was previously annotated with `@Ignore`, but it is passing. The comment that "rlike not supported by all datastores yet" should not be relevant, as this test runs exclusively in the `grails-datastore-gorm-test` project where it passes successfully.

matrei committed:
build: use `org.grails.grails-publish` for publishing

matrei committed:
fix(deps): bump `com.gradle.develocity` from `3.19` to `3.19.1`

- [3.19.1 Release Notes](https://plugins.gradle.org/plugin/com.gradle.develocity/3.19.1)

jdaugherty commented on .github/workflows/release.yml:
```diff
@@ -1,127 +1,126 @@
         if: success()
         uses: grails/github-actions/post-release@main
-        with:
-          token: ${{ secrets.GITHUB_TOKEN }}
```
Isn't the token so we can close out milestones, etc?  I'm ok leaving this out for now, but we probably should talk to the team how we want to handle closing out projects, etc.

matrei replied:
The token argument defaults to `github.token`, so we don't have to set it explicitly:
https://github.com/grails/github-actions/blob/15e3017e73c5c4bfdc02c36792da171418da563f/post-release/action.yml#L7
